ENH: Add qSlicerSingletonViewFactory

Sunderlandkyl opened the pull request:
The new view factory can be used by both C++ and Python classes to register any singleton QWidget by calling factory->setWidget(QWidget*) and factory->setTagName(std::string).

Since the widget is a singleton, the factory can only be used to add a single such view to a given layout.

Spun off from https://github.com/Slicer/Slicer/pull/1061

lassoan commented on Base/QTGUI/qSlicerSingletonViewFactory.cxx:
```diff
@@ -0,0 +1,122 @@
+
+  virtual void init();
+
+  QWidget* Widget;
```
Widget should be initialized to NULL in the constructor.

lassoan commented on Base/QTGUI/qSlicerSingletonViewFactory.h:
```diff
@@ -0,0 +1,68 @@
+/// This class provides an interface for C++ and Python classes to register a singleton view replacement widget.
+/// New view widgets can be registered by registering a new qSlicerSingletonViewFactory, and then setting the widget and tag using setWidget(QWidget*) and
+/// setTagName(QString).
+/// This factory contains a single pointer to an instance of the widget, so only one view can be created within a given layout.
```
What happens if two instances are used in a layout? Could we check this and report an error if it happens?

Sunderlandkyl replied:
If two or more of the same singleton view are included in the same layout, then only the last one will be filled. Previous views will be blank.

Probably could log an error if this happens.

Sunderlandkyl replied:
If two of the same views are created within the same layout, the following error message is now displayed:
````
qSlicerSingletonViewFactory::createViewFromXML - Widget for view "TagName" is already in use within the current layout!
````

lassoan commented on Base/QTGUI/qSlicerSingletonViewFactory.h:
```diff
@@ -0,0 +1,68 @@
+  virtual QStringList supportedElementNames()const;
+
+  /// Set the singleton widget instance that will be used to create the view
+  Q_INVOKABLE virtual void setWidget(QWidget* widget);
```
Who is responsible for deleting the widget? It should be at least clarified in the documentation. We may consider allowing the factory taking responsibility for deleting the widget if that simplifies the implementation.

Sunderlandkyl replied:
I changed the regular pointer to a QSharedPointer.
Factory is now responsible for deleting the widget.

lassoan reviewed (commented):
Thank you it looks good overall. There are just a few things to clarify - see inline comments.

Sunderlandkyl committed:
ENH: Add qSlicerSingletonViewFactory

The new view factory can be used by both C++ and Python classes to register any singleton QWidget by calling factory->setWidget(QWidget*) and factory->setTagName(std::string).
Since the widget is a singleton, a factory can only be used to add a single view to a given layout.

Sunderlandkyl commented:
Any other changes/comments?

lassoan commented:
Thank you, this looks good to me. I'll merge it now.

lassoan commented:
Merged, thank you!
